python/dgl/dataloading/dataloader.py

```diff
@@ -1,900 +1,602 @@
         else:
-            exclude_eids = F.copy_to(exclude_eids, device)
-
-    excluder = _EidExcluder(exclude_eids)
-    return subg if excluder is None else excluder(subg)
-
-
-def _find_exclude_eids_with_reverse_id(g, eids, reverse_eid_map):
-    if isinstance(eids, Mapping):
-        eids = {g.to_canonical_etype(k): v for k, v in eids.items()}
-        exclude_eids = {
-            k: F.cat([v, F.gather_row(reverse_eid_map[k], v)], 0)
-            for k, v in eids.items()}
+            meta_info = torch.LongTensor([0, 0])
+
+        if dist.get_backend() == 'nccl':
+            # Use default CUDA device; PyTorch DDP required the users to set the CUDA
+            # device for each process themselves so calling .cuda() should be safe.
+            meta_info = meta_info.cuda()
+        dist.broadcast(meta_info, src=0)
+
+        if self.rank != 0:
+            id_, num_samples = meta_info.tolist()
+            name = _get_shared_mem_name(id_)
+            if isinstance(indices, Mapping):
+                indices_shared = get_shared_mem_array(name, (num_samples, 2), torch.int64)
+            else:
+                indices_shared = get_shared_mem_array(name, (num_samples,), torch.int64)
+            self._tensor_dataset = indices_shared
+            self._device = indices_shared.device
+
+    def shuffle(self):
+        """Shuffles the dataset."""
+        # Only rank 0 does the actual shuffling.  The other ranks wait for it.
+        if self.rank == 0:
+            self._tensor_dataset[:self.num_indices] = self._tensor_dataset[
+                torch.randperm(self.num_indices, device=self._device)]
+            if not self.drop_last:
+                # pad extra
+                self._tensor_dataset[self.num_indices:] = \
+                    self._tensor_dataset[:self.total_size - self.num_indices]
+        dist.barrier()
+
+    def __iter__(self):
+        start = self.num_samples * self.rank
+        end = self.num_samples * (self.rank + 1)
+        dataset = _divide_by_worker(self._tensor_dataset[start:end])
+        return _TensorizedDatasetIter(
+            dataset, self.batch_size, self.drop_last, self._mapping_keys)
+
+    def __len__(self):
+        return (self.num_samples + (0 if self.drop_last else (self.batch_size - 1))) // \
+            self.batch_size
+
+
+def _prefetch_update_feats(feats, frames, types, get_storage_func, id_name, device, pin_memory):
+    for tid, frame in enumerate(frames):
+        type_ = types[tid]
+        default_id = frame.get(id_name, None)
+        for key in frame.keys():
+            column = frame[key]
+            if isinstance(column, LazyFeature):
+                parent_key = column.name or key
+                if column.id_ is None and default_id is None:
+                    raise DGLError(
+                        'Found a LazyFeature with no ID specified, '
+                        'and the graph does not have dgl.NID or dgl.EID columns')
```

Comment:
I wonder how to fix it if a user got such an error? Consider reformatting the error message as follows:
```python
Feature "{key}" is a LazyFeature but with no ID specified. Please create a LazyFeature with ....
```

Reply:
This only happens when people wants to prefetch features that are neither ndata nor edata.  In that case he/she needs to specify IDs in the LazyFeature object as well.